amalgamate.mjs

```diff
@@ -0,0 +1,264 @@
+                directory,
+                file,
+                relativePath,
+                output,
+                command,
+                arguments: args,
+                flags,
+                defines,
+                setCommonDefines,
+                flagsWithoutDefines,
+                includeLines,
+                namespace,
+            };
+        })
+        // Exclude unit tests as that compiles things twice, leading to redefinition errors.
+        .filter(e => !e.defines().find(define => define.name === "_UNIT_TEST_"))
+        .filter(e => !skippedFiles.has(e.relativePath))
+        ;
+
+    function intersection(...sets) {
+        const counter = new Map();
+        for (const set of sets) {
+            for (const e of set) {
+                counter.set(e, (counter.get(e) ?? 0) + 1);
+            }
+        }
+        return [...counter.entries()]
+            .filter(([e, count]) => count == sets.length)
+            .map(([e, count]) => e)
+            ;
+    }
+
+    // defines defined by every translation unit
+    const commonDefines = new Set(intersection(...commands.map(cmd => cmd.defines().map(define => define.flag))));
+
+    const flags = [
+        ...new Set(commands.flatMap(e => e.flagsWithoutDefines)),
+        ...commonDefines,
+        // "-ferror-limit=0",
+        "-Wno-unknown-warning-option",
+        `-I${directory}`,
+    ];
+
+    commands.forEach(cmd => {
+        cmd.setCommonDefines(commonDefines);
+    });
+
+    const includes = [
+        {
+            includeLines: () => [
+                // Need to define `_GNU_SOURCE` the first time we `#include` these.
+                "#define _GNU_SOURCE",
+                "#include <pthread.h>",
+                "#include <sched.h>",
+                "#undef _GNU_SOURCE",
+            ],
+        },
+        ...commands
+            .filter(cmd => !filesWithErrors.has(cmd.relativePath)),
+    ]
+        .map(cmd => cmd.includeLines().join("\n"))
+        .join("\n\n")
+        + "\n"
+        ;
+
+    const linkFlags = [];
+
+    const cc = process.env.CC ?? "cc";
+
+    const amalgamated = {
+        ii: {
+            // A post-preprocessing file, just to check exactly what's included or not.
+            directory,
+            arguments: [cc, ...flags, "-o", amalgamatedPath(".ii"), "-E", amalgamatedPath(".c")],
+            file: amalgamatedPath(".c"),
+        },
+        o: {
+            directory,
+            arguments: [cc, ...flags, "-o", amalgamatedPath(".o"), "-c", amalgamatedPath(".c")],
+            file: amalgamatedPath(".c"),
+        },
+        exe: {
+            directory,
+            arguments: [cc, ...flags, "-o", amalgamatedPath("exe"), amalgamatedPath("c"), ...linkFlags],
+            file: amalgamatedPath(".c"),
+        },
+    };
+
+    return {
+        amalgamated,
+        includes,
+    };
+}
+
+
+
+function toShellCommand(cmd) {
+    const command = cmd.arguments && cmd.arguments.join(" ") || cmd.command;
```

Comment:
Do we need to escape the arguments here?

Reply:
Probably?